File: aoc/2015/cpp/findthefloor.cpp
// An opening parenthesis, (, means he should go up one floor, 
// and a closing parenthesis, ), means he should go down one floor.
// 
// The apartment building is very tall, and the basement is very deep;
//
// To what floor do the instructions take Santa?

#include <iostream>
#include <fstream>
#include <string>

int main() {
    std::ifstream file("day1input.txt");
    std::string line;
    int floor {0};

    if (file.is_open()) {
        std::getline(file, line);
        if (line.size()) {
            for (const char& ch: line) {
                // increment the counter
                if (ch == '(') {
                    floor++;
                } else { // decrement the counter
                    floor--;
                }
            }

            std::cout<<"Santa reaches the floor "<<floor<<std::endl;
        } else {
            std::cout<<"File does not contain any data"<<std::endl;
        }
        file.close();
    } else {
        std::cout<<"Could not open the file"<<std::endl;
    }

    return 0;
}

File: aoc/2015/rust/findthefloor.rs
// An opening parenthesis, (, means he should go up one floor, 
// and a closing parenthesis, ), means he should go down one floor.
// 
// The apartment building is very tall, and the basement is very deep;
//
// To what floor do the instructions take Santa?

use std::fs::File;
use std::io::{self, Read};

fn main() -> io::Result<()> {
    let mut file = File::open("../day1input.txt")?;

    // Initialize a string to store the file contents
    let mut line = String::new();
    
    // Read the file contents into the string
    file.read_to_string(&mut line)?;

    let mut floor = 0;

    for ch in line.chars() {
        if ch == '(' {
            floor = floor + 1; // increment the counter
        } else {
            floor = floor - 1; // decrement the counter
        }
    }

    println!("Santa reaches the floor {}", floor);

    Ok(())
}
